### PR TITLE
Refactor `Commands` module

### PR DESCRIPTION
**Pull request by neonichu**
This is the initial step in reducing the first stage of the bootstrap build. Currently, the initial bootstrap build using CMake is essentially building the entirety of SwiftPM and that is (mostly) because the `Commands` module pulls in everything that SwiftPM possibly offers.

This introduces a new `CoreCommands` module which contains the necessary parts for providing a minimal build command using the native build system. This can be used to provide a new `swift-bootstrap` executable which is a minimal version of `swift-build`, but actually doing so will be a separate PR.

The Windows build currently relies entirely on CMake, so initally we will not be able to profit much from the minimal build, but this will get us closer to doing so in the future.

### CODE REVIEW

**Comment by tomerd** [comment]
very cool!

**Comment by neonichu** [comment]
@swift-ci please smoke test

**Comment by neonichu** [comment]
https://github.com/apple/sourcekit-lsp/pull/650
@swift-ci please smoke test

**Comment by neonichu** [comment]
https://github.com/apple/swift/pull/61574
https://github.com/apple/sourcekit-lsp/pull/650
@swift-ci please smoke test

**Comment by stevapple** [comment]
Bootstrapping SwiftPM in Windows with "debug" configuration is possible, but in release build it will hit apple/swift-llbuild#844, so this is currently unusable. It’s nice to see SwiftPM being bootstrapped on Windows, in align with other platforms.

**Comment by neonichu** [comment]
https://github.com/apple/swift/pull/61574
https://github.com/apple/sourcekit-lsp/pull/650
@swift-ci please smoke test macOS

**Comment by neonichu** [comment]
`fatal error encountered while reading from module 'CxxVectorSum'`

Looks like this is an unrelated issue on the CI right now? I also saw it on TSC PRs.

**Comment by neonichu** [comment]
https://github.com/apple/swift/pull/61574
@swift-ci please smoke test macOS

**Comment by neonichu** [comment]
Split out #5846 to address the concerns about build system support.

**Comment by neonichu** [comment]
Simplified this PR a little bit for easier review and updated it based on the refactoring I have done separately.

Once this lands, there will be two eventual follow-up PRs:
- a pretty simple PR that introduces the `swift-bootstrap` command and uses it for the `bootstrap` script
- a PR which changes the bootstrap process and CMake configuration to actually be more minimal. This requires us to move the Windows build to do bootstrapping first, so it'll take us a few steps to get there.

**Comment by neonichu** [comment]
@swift-ci please smoke test

**Comment by neonichu** [comment]
@swift-ci please smoke test

File: Sources/Commands/SwiftBuildTool.swift [hunk 1]
```diff
@@ -157,8 +159,23 @@ public struct SwiftBuildTool: SwiftCommand {
...
 
-extension Basics.Diagnostic {
-    static func mutuallyExclusiveArgumentsError(arguments: [String]) -> Self {
-        .error(arguments.map{ "'\($0)'" }.spm_localizedJoin(type: .conjunction) + " are mutually exclusive")
+extension SwiftCommand {
+    public func buildSystemProvider(_ swiftTool: SwiftTool) throws -> BuildSystemProvider {
+        return .init(providers: try swiftTool.defaultBuildSystemProvider.providers.merging([
+            .xcode: { (explicitProduct: String?, cacheBuildManifest: Bool, customBuildParameters: BuildParameters?, customPackageGraphLoader: (() throws -> PackageGraph)?, customOutputStream: OutputByteStream?, customLogLevel: Basics.Diagnostic.Severity?, customObservabilityScope: ObservabilityScope?) throws -> BuildSystem in
+                let graphLoader = { try swiftTool.loadPackageGraph(explicitProduct: explicitProduct) }
+                // FIXME: Implement the custom build command provider also.
```
Comment (tomerd) [inline Sources/Commands/SwiftBuildTool.swift:167]: still to do?

File: Sources/Commands/SwiftPackageTool.swift [hunk 2]
```diff
@@ -1190,7 +1191,7 @@ final class PluginDelegate: PluginInvocationDelegate {
         buildParameters.flags.linkerFlags.append(contentsOf: parameters.otherLinkerFlags)
 
         // Configure the verbosity of the output.
-        let logLevel: Diagnostic.Severity
+        let logLevel: Basics.Diagnostic.Severity
```
Comment (tomerd) [inline Sources/Commands/SwiftPackageTool.swift:1194]: needed?

File: Sources/Commands/Utilities/APIDigester.swift [hunk 3]
```diff
@@ -62,7 +63,7 @@ struct APIDigesterBaselineDumper {
         for modulesToDiff: Set<String>,
         at baselineDir: AbsolutePath?,
         force: Bool,
-        logLevel: Diagnostic.Severity,
+        logLevel: Basics.Diagnostic.Severity,
```
Comment (tomerd) [inline Sources/Commands/Utilities/APIDigester.swift:66]: needed?

File: Sources/Commands/SwiftBuildTool.swift [hunk 4]
```diff
@@ -157,8 +159,23 @@ public struct SwiftBuildTool: SwiftCommand {
...
+                return try XcodeBuildSystem(
+                    buildParameters: customBuildParameters ?? swiftTool.buildParameters(),
+                    packageGraphLoader: customPackageGraphLoader ?? graphLoader,
+                    outputStream: customOutputStream ?? swiftTool.outputStream,
+                    logLevel: customLogLevel ?? swiftTool.logLevel,
+                    fileSystem: swiftTool.fileSystem,
+                    observabilityScope: customObservabilityScope ?? swiftTool.observabilityScope
+                )
+            },
+        ], uniquingKeysWith: { a, b in
```
Comment (tomerd) [inline Sources/Commands/SwiftBuildTool.swift:177]: this is a bit unclear to me, we are merging a map of 1?

File: Sources/CoreCommands/MinimalBuildTool.swift [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+    }
+
+    func didDownloadBinaryArtifact(from url: String, result: Result<AbsolutePath, Error>, duration: DispatchTimeInterval) {
+    }
+
+    func downloadingBinaryArtifact(from url: String, bytesDownloaded: Int64, totalBytesToDownload: Int64?) {
+    }
+
+    func didDownloadAllBinaryArtifacts() {
+    }
```
Comment (tomerd) [inline Sources/CoreCommands/MinimalBuildTool.swift:87]: nit: make all of these into one liners `{}`

**Review by tomerd — APPROVED** [review]
very nice

File: Sources/Commands/SwiftBuildTool.swift (hunk 1 above)
Comment (neonichu) [reply Sources/Commands/SwiftBuildTool.swift:167]: I just carried this over from the original code, honestly not sure what it means, so maybe we can remove it?

File: Sources/Commands/SwiftPackageTool.swift (hunk 2 above)
Comment (neonichu) [reply Sources/Commands/SwiftPackageTool.swift:1194]: Yep

File: Sources/Commands/Utilities/APIDigester.swift (hunk 3 above)
Comment (neonichu) [reply Sources/Commands/Utilities/APIDigester.swift:66]: Yep

File: Sources/Commands/SwiftBuildTool.swift (hunk 4 above)
Comment (neonichu) [reply Sources/Commands/SwiftBuildTool.swift:177]: We're merging with the providers from `defaultBuildSystemProvider` above, right now this doesn't have an effect, but I wanted to keep it forward compatible.

File: Sources/Commands/SwiftBuildTool.swift (hunk 1 above)
Comment (neonichu) [reply Sources/Commands/SwiftBuildTool.swift:167]: Removing this now.

**Comment by neonichu** [comment]
@swift-ci please smoke test

**Pull request merged**